pkg/ccl/backupccl/restore_data_processor.go

```diff
@@ -296,18 +354,69 @@ func (rd *restoreDataProcessor) openSSTs(
 	log.VEventf(ctx, 1 /* level */, "ingesting span [%s-%s)", entry.Span.Key, entry.Span.EndKey)
 
 	storeFiles := make([]storageccl.StoreFile, 0, len(entry.Files))
-	for _, file := range entry.Files {
+	iterAllocs := make([]*quotapool.IntAlloc, 0, len(entry.Files))
+	var sstOverheadBytesPerFile uint64
+	if rd.spec.Encryption != nil {
+		sstOverheadBytesPerFile = sstReaderEncryptedOverheadBytesPerFile
+	} else {
+		sstOverheadBytesPerFile = sstReaderOverheadBytesPerFile
+	}
+
+	for idx := resumeIdx; idx < len(entry.Files); {
+		file := entry.Files[idx]
 		log.VEventf(ctx, 2, "import file %s which starts at %s", file.Path, entry.Span.Key)
 
+		alloc, err := rd.sstMemQuotaPool.TryAcquire(ctx, sstOverheadBytesPerFile)
+		if errors.Is(err, quotapool.ErrNotEnoughQuota) {
+			if err := rd.mem.Grow(ctx, int64(sstOverheadBytesPerFile)); err != nil {
+				// If we failed to allocate more memory, send the iterator
+				// containing the files we have right now.
+				if len(storeFiles) > 0 {
+					iterOpts := storage.IterOptions{
+						RangeKeyMaskingBelow: rd.spec.RestoreTime,
+						KeyTypes:             storage.IterKeyTypePointsAndRanges,
+						LowerBound:           keys.LocalMax,
+						UpperBound:           keys.MaxKey,
+					}
+					iter, err := storageccl.ExternalSSTReader(ctx, storeFiles, rd.spec.Encryption, iterOpts)
+					if err != nil {
+						return mergedSST{}, 0, err
+					}
+
+					log.VInfof(ctx, 2, "sending iterator after %d out of %d files due to insufficient memory", idx, len(entry.Files))
+
+					mSST, err := getIter(iter, dirs, iterAllocs, false)
+					return mSST, idx, err
+				}
+
+				alloc, err = rd.sstMemQuotaPool.Acquire(ctx, sstOverheadBytesPerFile)
```

Comment:
so, the following can happen:
- thread 1 does `TryAcquire` and fails
- thread 1 `Grow`s the memory reservation
- thread 2 does `TryAcquire` and succeeds
- thread 1 get here to `Acquire` and hangs.
perhaps it's not a sever issue like a deadlock but if the memory monitor and the quota pool are abstracted away as a single object we can avoid those kind of issues, and for example in this case `Grow` and `TryAcquire` in a loop until we acquired or failed without hanging.

Reply:
added abstraction